storage/filestore.go

```diff
@@ -61,6 +67,73 @@ type FilesystemStore struct {
 	perms   os.FileMode
 }
 
+func (f *FilesystemStore) moveKeyTo0Dot4Location(file string) {
+	keyID := filepath.Base(file)
+	fileDir := filepath.Dir(file)
+	d, _ := f.Get(file)
+	block, _ := pem.Decode(d)
+	if block == nil {
+		logrus.Warn("Key data for", file, "could not be decoded as a valid PEM block. The key will not been migrated and may not be available")
+		return
+	}
+	fileDir = strings.TrimPrefix(fileDir, notary.RootKeysSubdir)
+	fileDir = strings.TrimPrefix(fileDir, notary.NonRootKeysSubdir)
+	if fileDir != "" {
+		block.Headers["gun"] = fileDir[1:]
+	}
+	if strings.Contains(keyID, "_") {
+		role := strings.Split(keyID, "_")[1]
+		keyID = strings.TrimSuffix(keyID, "_"+role)
+		block.Headers["role"] = role
+	}
+	var keyPEM bytes.Buffer
+	// since block came from decoding the PEM bytes in the first place, and all we're doing is adding some headers we ignore the possibility of an error while encoding the block
+	pem.Encode(&keyPEM, block)
+	f.Set(keyID, keyPEM.Bytes())
+}
+
+func (f *FilesystemStore) migrateTo0Dot4() {
+	rootKeysSubDir := filepath.Join(f.Location(), notary.RootKeysSubdir)
+	nonRootKeysSubDir := filepath.Join(f.Location(), notary.NonRootKeysSubdir)
+	if _, err := os.Stat(rootKeysSubDir); !os.IsNotExist(err) && f.Location() != rootKeysSubDir {
+		// root_keys exists, migrate things from it
+		listOnlyRootKeysDirStore, _ := NewFileStore(rootKeysSubDir, f.ext, notary.PrivKeyPerms)
+		for _, file := range listOnlyRootKeysDirStore.ListFiles() {
+			f.moveKeyTo0Dot4Location(filepath.Join(notary.RootKeysSubdir, file))
+		}
+		// delete the old directory
+		if rootKeysSubDir == "" || rootKeysSubDir == "/" {
```

Comment:
@cyli is 100% correct. I asked @avaid96 to put these checks in as a sanity check in case anyone is making updates in the future and does something silly.

@cyli has also made the suggestion that we check `f.Location` and the `(Non)RootKeysSubdir` constants at the start instead of checking the joined values here and later. That would stop a possible walk over `/` which would take a long time.